### ISSUE TITLE
Add type checks for `assert.Regexp` regexp argument

### ISSUE THREAD

**Issue by dolmen** [Antonboom/testifylint#81]
The signature for [`assert.Regexp`](https://pkg.go.dev/github.com/stretchr/testify/assert#Regexp) (and associated methods) is:

```go
func Regexp(t TestingT, rx interface{}, str interface{}, msgAndArgs ...interface{}) bool
```

However `rx` should only be `*regexp.Regexp` or `string`.

Related: stretchr/testify#1585
Cc: @kevinburkesegment

**Comment by ccoVeille** [comment]
It's interesting.

Could you explain the unexpected behavior you found with this method, especially the "around fmt.Sprint()"

Do you have examples of interface value passed to assert.Regexp ?

**Comment by dolmen** [comment]
I just tried a [code search on GitHub](https://github.com/search?q=assert.NotRegexp+language%3AGo+&type=code&p=4) of bad usage of `assert.NotRegexp`, and found none so far, so bad usage is probably quite rare and usually easily caught by running the test.

But that would be the point of this checker: catch the remaining ones.

About `fmt.Sprint()` traps: conversion of `[]byte` is often not what one expects. See https://go.dev/play/p/3XsaeWgDsUD

**Comment by ccoVeille** [comment]
Side note, if something is implemented it cannot about detecting the value is a regexp object or a string. The code use fmt.Sprint, so it relies also in being able to convert into a string.

So it should also support fmt.Stringer

https://github.com/golang/go/blob/master/src%2Ffmt%2Fprint.go

**Comment by dolmen** [comment]
We are discussing about a linter.

As a Testify maintainer I can't change the crazy things that are allowed by the current implementation. But I expect the linter to tell the user that the crazy thing he does is most probably just a mistake.

**Comment by ccoVeille** [comment]
Yes, I agree that's why I listed the things the linter could raise.

I agree with you, it's not about changing testify, but to guide user to use the right asserter

**Comment by ccoVeille** [comment]
> @ccoVeille We are discussing about a linter.
> 
> As a Testify maintainer I can't change the crazy things that are allowed by the current implementation. But I expect the linter to tell the user that the crazy thing he does is most probably just a mistake.

About this, because I'm unsure about your reply.

Your issue was about accepting:
- string
- *regexp.Regexp

I suggested current implementation also support any struct with a stringer.

So something like this:

```go
type alpha struct{}

func (*alpha) String() string {
  return "[a-z]+"
}

```
(This is pseudo code made on a phone)

Would you expect the linter to report sending any object from this type to `assert.Regexp` as invalid?

**Comment by Antonboom** [comment]
\+ case when order of arguments is invalid

```go
assert.Regexp(t, out, `\[.*\] DEBUG \(.*TestNew.*\): message`)
```

**Comment by Antonboom** [comment]
+ https://github.com/stretchr/testify/pull/1587

**Comment by brackendawson** [comment]
> what are the reasons to support everything else?

Quite simply that testify already does, to change it is a breaking change.

I agree that it's utter insanity, but the documentation for assert.Regexp should be changed to explain that if rx is not a *regexp.Regexp then the `%v` format of rx will be compiled as a regexp. I'll raise a new issue for that.

Also this panics:
```go
func TestIfy(t *testing.T) {
	require.Regexp(t, "\\C", "I don't like this at all")
}
```
I'll raise a new issue for that too.

**Comment by brackendawson** [comment]
I think it would be sensible to consider any rx type other than *regexp.Regexp or string to be a lint error.

**Comment by Antonboom** [comment]
But failing in `IsIncreasing` is a breaking change too? :trollface: 

Maybe soften the requirements for absolutely meaningless and product-damaging cases?

**Comment by brackendawson** [comment]
Sure, every bug fix is a breaking change if you look at it closely enough. For cases where a passing test will turn into a failing test I draw the line at whether the test should fail for what the unsuspecting user wrote. Make a test fail when it doesn't test what the user thought it did, but don't break a test when it does test what the user thought it does just because I find the API distasteful.

eg:
I can only think of cases where IsIncreasing was used wrong as being mistakes, such missing a colon in a re-slice or typing the wrong variable name. The user thinks they tested that their slice/array increases but actually no such assertion is made, we should have highlighted the mis-use. The by-catch would be people who thought other types can be increasing, like string, but again we allowed them to think they asserted something that we don't. It also breaks the invariant where assertions return false when they fail the test and true when they do not. This lands in the bug category for me.
```go
assert.IsIncreasing(t, s[1]) // maybe should have been s[1:]
```

For regexp; I find this abhorrent, it might be the user forgot to type the quotes, or it might have been intentional, but testify is making the assertion that the user wanted. This lands in the feature category for me. I still hate it, if testify were to ever have a v2 I would remove this. But banning this is a breaking change.
```go
assert.Rexexp(t, true, "too true")
```

**Comment by ccoVeille** [comment]
Some interesting feedback from @dolmen made on #292

> We must only accept:
> * any `*types.TypeParam`
> * `*types.Basic` that is strict `string` (BasicKind is `String` or `UntypedString`) (but NOT anything that is `string` underlying)
> * any `*types.Alias` to above
> 
> The current implementation is too lax and doesn't handle TypeParam.
> 
> _Originally posted by @dolmen in https://github.com/Antonboom/testifylint/pull/292#pullrequestreview-4352355204_
